src/unity/toolkits/supervised_learning/supervised_learning.cpp

```diff
@@ -1107,46 +1113,58 @@ gl_sframe supervised_learning_model_base::api_classify(
  *  Evaluate the model
  */
 variant_map_type supervised_learning_model_base::api_evaluate(
-    gl_sframe data, std::string missing_value_action_str, std::string metric) {
+    gl_sframe data, std::string missing_value_action_str, std::string metric, gl_sarray predictions) {
   auto model = std::dynamic_pointer_cast<supervised_learning_model_base>(
       shared_from_this());
   ml_missing_value_action missing_value_action =
       get_missing_value_enum_from_string(missing_value_action_str);
 
-  sframe test_data = data.materialize_to_sframe();
-  sframe X = setup_test_data_sframe(test_data, model, missing_value_action);
-  sframe y = test_data.select_columns({get_target_name()});
-  ml_data m_data = setup_ml_data_for_evaluation(
-      X, y, model, missing_value_action);
-
+  variant_map_type ret;
 
-  if(metric == "report") {
+  if(metric == "report" || metric == "auto") {
```

Comment:
I found all this logic very difficult to follow.

AFAICT, the original logic basically said:
- If the requested metric is "report" and the model is a classifier, then just return the confusion matrix, report by class, and accuracy.
- If the requested metric is "report" and the model is not a classifier, then treat "report" as "auto
- For non-"report" metrics, just call the underlying `evaluate` method

The new logic basically says:
1. If the metric is "report" or "auto", and the model is a classifier, then compute the confusion matrix, report by class, and accuracy
2. If the metric is NOT "report" OR if the model is NOT a classifier, then compute the results of calling `evaluate`
3. Return the combined the results, if any, from 1 and 2 above

Is this the intent? (That is, add confusion matrix, report by class, and accuracy to all classification toolkits by default?) If so, could we make the logic more like the above? Just create a single accumulator map for all the results. If desired, call out to one function to add the classification reports. Then if necessary, call out to `evaluate` to add the traditional stuff?

Reply:
Yes, I agree.  This logic is an evolutionary walk of patches to get stuff to work.  I think it should be completely refactored at this point.  Let me do that.

Reply:
In my other comment, I am especially referring to the logic here